### ISSUE TITLE
Truncate with foreign key check ignore

### ISSUE THREAD

**Issue by Soullivaneuh** [hautelook/AliceBundle#121]
When `--purge-with-truncate` options is used, foreign key check should be deactivated.

If not, most of projects will get this kind of error:

```
SQLSTATE[42000]: Syntax error or access violation: 1701 Cannot truncate a table referenced in a foreign key constraint
```

**Comment by theofidry** [comment]
@Soullivaneuh is the bug still present? There was an error in a recent version which was ignoring the truncate option.

**Comment by Soullivaneuh** [comment]
Got the error with following packages:

```bash
$ composer info -i | grep alice
hautelook/alice-bundle                   v1.0.0-beta.5       Symfony2 Bundle to manage fixtures with Alice and Faker.
nelmio/alice                             2.1.0               Expressive fixtures generator
```

**Comment by Soullivaneuh** [comment]
Seems to be related to doctrine data-fixtures:

* https://github.com/doctrine/data-fixtures/issues/113
* https://github.com/doctrine/data-fixtures/pull/127

**Comment by theofidry** [comment]
Looks like quite the nasty issue there... I will remove it from the v1.0.0 milestone as it's to be handled in [Doctrine data fixtures extension](https://github.com/doctrine/data-fixtures)

**Comment by theofidry** [comment]
@Soullivaneuh what do you think of doing it [this way](https://github.com/doctrine/data-fixtures/issues/113#issuecomment-144950542)?

**Comment by Soullivaneuh** [comment]
I saw that, but didn't test yet. But yes, could be a workaround.

I said a workaround, not a solution. :-)

**Comment by theofidry** [comment]
There is also [LaFourchetteFixturesBundle](https://github.com/lafourchette/FixturesBundle) that apparently found a solution, didn't have time to check how yet.

**Comment by Soullivaneuh** [comment]
Ha ha ha ha! https://github.com/doctrine/dbal/blob/f44782e91c8c910736272986f543989d214daf1c/lib/Doctrine/DBAL/Platforms/AbstractPlatform.php#L3413

Unused variable `$cascade`... loving it. :-')

**Comment by theofidry** [comment]
:skull:

**Comment by Soullivaneuh** [comment]
This workaround works great, indeed.

Don't forget to set `config_class` to null on non-mysql connections (for config_test with sqlite db for example).

**Comment by vincentchalamon** [comment]
@theofidry [LaFourchetteFixturesBundle doesn't care about other platforms than MySQL](https://github.com/lafourchette/FixturesBundle/blob/master/Loader/FixturesLoader.php#L104). I'm not even sure this request could work on other platforms than MySQL…

I remember DoctrineFixturesBundle has the same issue since a loooong time.

**Comment by theofidry** [comment]
Arf, such a pain...

**Comment by vincentchalamon** [comment]
Couldn't it be done in [FixturesExecutor](https://github.com/hautelook/AliceBundle/blob/1.x/src/Doctrine/DataFixtures/Executor/FixturesExecutor.php) ?

**Comment by theofidry** [comment]
@vincentchalamon so you would check the platform and if is MySQL unset the foreign key check?

**Comment by vincentchalamon** [comment]
@theofidry Isn't it what it's already done in FixturesExecutor ?
```php
        switch (true) {
            case $manager instanceof EntityManagerInterface:
                $executor = new ORMExecutor($manager, $loader);
                $purger = new ORMPurger($manager);
                $purger->setPurgeMode(
                    $purgeMode
                        ? ORMPurger::PURGE_MODE_TRUNCATE
                        : ORMPurger::PURGE_MODE_DELETE
                );
                break;
            case $manager instanceof MongoDBDocumentManager:
                $executor = new MongoDBExecutor($manager, $loader);
                $purger = new MongoDBPurger($manager);
                break;
            case $manager instanceof PHPCRDocumentManager:
                $executor = new PHPCRExecutor($manager, $loader);
                $purger = new PHPCRPurger($manager);
                break;
            default:
                throw new \InvalidArgumentException(sprintf(
                    'Unsupported manager type %s',
                    get_class($manager))
                );
        }
```

**Comment by theofidry** [comment]
Nope, or at least according to [your example](https://github.com/lafourchette/FixturesBundle/blob/master/Loader/FixturesLoader.php#L104) you unset this check besides configuring the purger properly.

**Comment by vincentchalamon** [comment]
```$connection->getDriver()->getDatabasePlatform()->supportsForeignKeyConstraints();``` always returns true, except for SQLite which returns false.
But I'm more afraid about the query `SET FOREIGN_KEY_CHECKS = 0;` which can fail on some ORM systems :/ What's your opinion about this one ?

**Comment by theofidry** [comment]
If it works for MySQL only I would be ok to do it. We can know if we are dealing with a MySQL database right? It would be more like a hard check as apparently `supportsForeignKeyConstraints()` is unreliable...

**Comment by vincentchalamon** [comment]
You would like to check platform by platform ? Like for MySQL:
```php
$connection = $manager->getConnection();
if ($connection->getDriver()->getDatabasePlatform() instanceof MySqlPlatform) {
    $connection->exec('SET FOREIGN_KEY_CHECKS = 0;');
}
// ...ORM purger -> execute
if ($connection->getDriver()->getDatabasePlatform() instanceof MySqlPlatform) {
    $connection->exec('SET FOREIGN_KEY_CHECKS = 1;');
}
```

**Comment by theofidry** [comment]
Yep, assuming it works for "all" MySQL versions..., you could take care of doing the PR?

**Comment by vincentchalamon** [comment]
Apparently, this issue occurred from MySQL 5.5. Is there a way to configure travis with different versions of MySQL (like it does for PHP) ?

I'm creating the PR. And it should be interesting to upgrade it in the future with other platforms (through other PRs, of course).

**Comment by theofidry** [comment]
No idea :( I guess there is some ways to do it but may be a bit hacky :/

**Comment by vladimirkr** [comment]
This bug steel appears.
When I execute: php bin/console hautelook:fixtures:load --purge-with-truncate
- no errors appears
- command executes with PurgeMode::createDeleteMode instead of PurgeMode::createTruncateMode

If I change string vendor/hautelook/alice-bundle/src/Loader/DoctrineOrmLoader.php:174:
return $loader->load($files, $parameters);
to:
return $loader->load($files, $parameters, [], PurgeMode::createTruncateMode());

I receive: An exception occurred while executing 'TRUNCATE blablabla':               
  SQLSTATE[42000]: Syntax error or access violation...

**Comment by theofidry** [comment]
This should already be the case: https://github.com/theofidry/AliceDataFixtures/blob/master/src/Bridge/Doctrine/Purger/Purger.php#L98

So either the dependencies are not up to date or the bug is something else.

**Comment by theofidry** [comment]
Well I guess it doesn't hurt to open the issue until guaranteed that it works fine

**Comment by flaushi** [comment]
Yeah, ok, sorry my mistake.
Have you tested this code on Postgres as well? Because AFAIK, the command `SET FOREIGN_KEY_CHECKS = 1;` is MySQL-specific...

**Comment by theofidry** [comment]
I don't think there is a test on PostgreSQL in AliceDataFixtures so it might be worth adding one indeed
